### PR TITLE
(#139) Creating new playlist

### PR DESCRIPTION
**Pull request by herrhotzenplotz**
This is not yet finished, though I appreciate some early feedback on this :smiley:

### CODE REVIEW

**Comment by herrhotzenplotz** [comment]
I guess this is finished and ready for review.

One follow-up issue/PR might be to add support for deleting/unfollowing playlists.

Fixes #139

File: lib/src/spotifyapi/playlists.cpp [hunk 1]
```diff
@@ -10,6 +10,18 @@ using namespace lib::spt;
 // playlists/{playlist_id}/tracks
 // playlists/{playlist_id}/images
 
+void api::create_playlist(const std::string &name,
+	const std::string &description, const bool is_public,
+	const bool is_collaborative, lib::callback<lib::spt::playlist> &callback)
+{
+	post(lib::fmt::format("me/playlists"), {
```
Comment (kraxarn) [inline lib/src/spotifyapi/playlists.cpp:17]: No need for a `lib::fmt::format` call here.

File: src/dialog/createplaylistdialog.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
+#include "createplaylistdialog.hpp"
+
+#include <QFormLayout>
+#include <QDialogButtonBox>
+#include <QPushButton>
+#include <QMessageBox>
+
+CreatePlaylistDialog::CreatePlaylistDialog(QWidget *parent)
```
Comment (kraxarn) [inline src/dialog/createplaylistdialog.cpp:8]: I would much prefer to try to merge this with the dialog for editing playlists. Or maybe even better, try to do it like the official client does, that is, only lets you choose a name for the playlist before creating. Other stuff like public, description, etc. have to be changed after creation.

File: src/dialog/createplaylistdialog.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	: QDialog(parent)
+{
+	auto *layout = new QVBoxLayout();
+	setLayout(layout);
+	setWindowTitle("Create new playlist");
+
+	auto *formsWidget = new QWidget;
```
Comment (kraxarn) [inline src/dialog/createplaylistdialog.cpp:15]: Prefer using new with (), that is `new QWidget()`.

File: src/dialog/createplaylistdialog.hpp [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+public:
+	CreatePlaylistDialog(QWidget *parent);
+
+	auto playlistName() -> std::string;
+	auto playlistDescription() -> std::string;
+	auto playlistPublic() -> bool;
+	auto playlistCollaborative() -> bool;
+
+private:
+	QLineEdit *name_editfield;
```
Comment (kraxarn) [inline src/dialog/createplaylistdialog.hpp:20]: Everything in `src/` more follows the style of Qt, camel case, so this should for example be `nameEditfield`.

File: src/menu/songmenu.cpp [hunk 5]
```diff
@@ -313,3 +285,73 @@ auto SongMenu::getTrackUrl() const -> QString
 	auto str = lib::fmt::format("https://open.spotify.com/track/{}", trackUri);
 	return QString::fromStdString(str);
 }
+
+void SongMenu::addToNewPlaylist()
+{
+	CreatePlaylistDialog create_playlist_dialog {this};
+	create_playlist_dialog.exec();
+
+	if (create_playlist_dialog.result() == QDialog::Rejected)
```
Comment (kraxarn) [inline src/menu/songmenu.cpp:294]: Missing brackets

**Review by kraxarn — CHANGES_REQUESTED** [review]
Thanks for your contribution!
Overall it looks good. I'm not sure if the dialog for creating a new playlist is really needed though, considering there already is one for editing.

File: src/dialog/createplaylistdialog.cpp (hunk 2 above)
Comment (herrhotzenplotz) [reply src/dialog/createplaylistdialog.cpp:8]: By that you mean, spawn a QInputDialog for entering the name and add a single track to the playlist?
That would then eliminate the need of the separate dialog and cohere with the behavior of the official client.

**Comment by herrhotzenplotz** [comment]
Pushed changes, now using a QInputDialog. That makes it much simpler. If that is not what you intended, I can certainly revert these changes.

For completeness, I left the API call for `api::create_playlist` with all its options. Unfortunately we cannot make them use default arguments because C++... however, since the call is not used anywhere else I could also just remove them.

If needed, I can squash the changes, such that the CreatePlaylistDialog does not appear in the history.

**Comment by kraxarn** [comment]
Something like that, although, the other issues need to be resolved first before I merge this.

**Review by kraxarn — APPROVED** [review]
Something like that, looks good now. Optional parameters are possible with for example pointers, but it's probably overkill for now. I'm ready to merge this now, and again, thanks for your contribution!

**Pull request merged**